### PR TITLE
Added progressive jpeg variation option

### PR DESCRIPTION
**Pull request by KostyaEsmukov**
See #100

### CODE REVIEW

File: tests/test_models.py [hunk 1]
```diff
@@ -140,13 +140,52 @@ def test_delete_thumbnail(self):
         path = os.path.join(IMG_DIR, 'image.thumbnail.gif')
         assert not os.path.exists(path)
 
-    def test_fore_min_size(self):
+    def test_force_min_size(self):
```
Comment (KostyaEsmukov) [inline tests/test_models.py]: This is a non-related typo fix, but I hope it's OK

File: tests/test_utils.py [hunk 2]
```diff
@@ -24,7 +24,9 @@ def test_render_variations(self, image_upload_file):
                     'width': 150,
                     'height': 150,
                     'crop': True,
-                    'resample': Image.ANTIALIAS
+                    'resample': Image.ANTIALIAS,
+                    'is_progressive_jpeg': False,
+                    'progressive_jpeg_quality': 80
```
Comment (codingjoe) [inline tests/test_utils.py]: please add a trailing comma

File: stdimage/models.py [hunk 3]
```diff
@@ -231,7 +251,7 @@ def set_variations(self, instance=None, **kwargs):
                 for name, variation in list(self.variations.items()):
                     variation_name = self.attr_class.get_variation_name(
                         field.name,
-                        variation['name']
+                        variation
```
Comment (codingjoe) [inline stdimage/models.py]: why that change?

Comment (codingjoe) [inline stdimage/models.py]: its unrelated

File: stdimage/models.py [hunk 4]
```diff
@@ -158,7 +174,9 @@ class StdImageField(ImageField):
         'width': float('inf'),
         'height': float('inf'),
         'crop': False,
-        'resample': Image.ANTIALIAS
+        'resample': Image.ANTIALIAS,
+        'is_progressive_jpeg': False,
+        'progressive_jpeg_quality': 80
```
Comment (codingjoe) [inline stdimage/models.py]: please add trailing comma

File: stdimage/models.py [hunk 5]
```diff
@@ -168,7 +186,9 @@ def __init__(self, verbose_name=None, name=None, variations=None,
         Standardized ImageField for Django.
 
         Usage: StdImageField(upload_to='PATH',
-         variations={'thumbnail': {"width", "height", "crop", "resample"}})
+         variations={'thumbnail': {"width", "height", "crop", "resample",
```
Comment (codingjoe) [inline stdimage/models.py]: I don't know if you even need to include this information at this point.

File: stdimage/models.py [hunk 6]
```diff
@@ -128,7 +144,7 @@ def delete(self, save=True):
         super(StdImageFieldFile, self).delete(save)
 
     def delete_variations(self):
-        for variation in self.field.variations:
+        for _, variation in self.field.variations.items():
```
Comment (codingjoe) [inline stdimage/models.py]: Why did you change that?

**Comment by codingjoe** [comment]
I didn't want to get this feature into the main API since it's file type specific. But there are many people who'd like to see this feature. I still don't know if it is a good idea to add it to the field signature.
I'd rather have a callable for jpegs that adds those options.

File: stdimage/models.py (hunk 6 above)
Comment (KostyaEsmukov) [inline stdimage/models.py]: I had to make the `StdImageFieldFile.get_variation_name` method use the full variation dict, because resulting filename now depends on more than just a variation name - the `is_progressive_jpeg` property might change filename extension. I've preserved ability to get filename by just a name, but that is error-prone and should not be used (i.e. filenames generated for progressive variations might have wrong extension).

I guess that ability of getting variation filename by just a variation name should be marked for deprecation.

File: stdimage/models.py (hunk 3 above)
Comment (KostyaEsmukov) [inline stdimage/models.py]: See my comment about the `StdImageFieldFile.get_variation_name` above.

**Comment by KostyaEsmukov** [comment]
Could you be more specific on that callable approach?

Please note that this is not file type specific. Progressive JPEG variations are generated for all image types, not just for a JPEG.

**Comment by codingjoe** [comment]
Hi @KostyaEsmukov,

there currently is the `render_variations` callable. Which allows you to bypass the current image processor and define your own one. Maybe it makes sense to extend this idea. To have a callable that allows you to change the image processing.

eg.
```
class MyRenderer:

    def __init__(self):
        pass

    def __call__(self, file_name, variations, storage):
        pass

   def render_variations(self, variation):
        pass

    def render_variation(self, variation):
        pass
```

I guess, having a good template here, would allow it to easily extend or change behavior.

**Comment by KostyaEsmukov** [comment]
Hi @codingjoe,

Custom renderers idea looks good to me. 

Variation filename generation should also be overridable here. For example, for a progressive jpeg renderer we need to force a progressive jpeg variation filename to have the `.jpg` extension.

But it should be borne in mind that once that custom variation filename generator is changed, paths to some already stored variations might be lost, thus loosing ability to remove them from a storage. I think it's a minor issue, but still. Although this behaviour is similar to changing the `variations` dict currently.

**Comment by codingjoe** [comment]
@KostyaEsmukov changing the way file names for variations are created, is something one should change without fully understanding the effects. I think a proper documentation should be enough tho.

**Comment by KostyaEsmukov** [comment]
Hi @codingjoe,

Did you come to a decision? How should we proceed?

**Comment by codingjoe** [comment]
@KostyaEsmukov I don't know, feel free to come up with a design as a base of discussion

**Comment by wiesson** [comment]
Any updates to this topic? I would like to post-process / replace the original the images with https://imageoptim.com/api (like the approach from @KostyaEsmukov) to create progressive jpgs and also reduce the file size. 

> I guess, having a good template here, would allow it to easily extend or change behavior.

I would be great, if you provide a template and a simple example how to start.

**Comment by codingjoe** [comment]
@KostyaEsmukov after reading [this great article]. I've come to the conclusion, that we should add progressive processing for pictures over 100x100 pixel by default.
That should solve you problem for now. This library is supposed to do the common case the best way possible. All special cases can be solved using inheritance or with a custom processor.

**Comment by KostyaEsmukov** [comment]
Unfortunately that doesn't solve my problem, because progressive JPEG variations are not generated for PNG, BMP, etc.

The problem here is variation file name handling -- it is assumed currently that variation file type must be the same as for the original image, that is not the case if I want to generate progressive JPEG variations for image types other than JPEG.

I tried to find a convenient way to extend the idea of own processor/custom renderer to handle this but it either looks awkward or results in a heavy refactoring.

**Comment by codingjoe** [comment]
@KostyaEsmukov hm... you could use the `upload_to` to transform the image into a JPEG.
If you want to keep the original but have all thumbnails as a JPEG that requires making multiple changes.

We talked about how we could improve the design to make this more easy. I'm open for suggestions. I'm afraid I have currently not the time nor the incentive to make such improvements myself.

**Comment by KostyaEsmukov** [comment]
Transformation to the JPEG is a good point indeed. Now that JPEG variations are progressive by default, it might work out for me. 

I will let you know if I come up with a good solution to my problem other than storing all images as JPEGs.

Thank you, @codingjoe
